### PR TITLE
feat: typed table features

### PR DESCRIPTION
**Pull request by roeap**
delta-kernel aiming to be a reference implementation of the protocol, I thought it might make sense to be explicit about what table features we know about.

This has also proven quite useful in delta-rs when implementing various features.

### CODE REVIEW

File: kernel/src/actions/mod.rs [hunk 1]
```diff
@@ -100,10 +254,12 @@ pub struct Protocol {
...
     /// A collection of features that a client must implement in order to correctly
     /// read this table (exist only when minReaderVersion is set to 3)
-    pub reader_features: Option<Vec<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub reader_features: Option<HashSet<String>>,
     /// A collection of features that a client must implement in order to correctly
     /// write this table (exist only when minWriterVersion is set to 7)
-    pub writer_features: Option<Vec<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub writer_features: Option<HashSet<String>>,
```
Comment (roeap) [inline kernel/src/actions/mod.rs:262]: Since probably not too critical I was on the fence about changing this, but then I though since we will mostly be checking if a feature exists, might as well and hear what others have to say.

Comment (nicklan) [reply kernel/src/actions/mod.rs:262]: I'd vote for leaving it as a `Vec`. Generally for  small numbers of items in the collection, a scan of the vec is faster.

Some data [here](https://gist.github.com/daboross/976978d8200caf86e02acb6805961195) although the test is slightly flawed. But for low numbers of elements `Vec` should generally be better.

File: kernel/src/actions/mod.rs [hunk 2]
```diff
@@ -90,7 +91,160 @@ impl Metadata {
     }
 }
 
-#[derive(Default, Debug, Clone, PartialEq, Eq, Schema)]
+/// Features table readers can support as well as let users know
```
Comment (nicklan) [inline kernel/src/actions/mod.rs:94]: can we move these into a `features` mod? In general I think it would be best to keep just the actual "Actions" in this mod (although it's probably already not perfect on that front)

**Review by nicklan — CHANGES_REQUESTED** [review]
mostly fine, just a couple of small suggestions.

File: kernel/src/actions/mod.rs [hunk 3]
```diff
@@ -90,7 +91,160 @@ impl Metadata {
...
+
+impl TryFrom<String> for ReaderFeatures {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.as_str().try_into()
+    }
+}
+
+impl TryFrom<&str> for ReaderFeatures {
```
Comment (scovich) [inline kernel/src/actions/mod.rs:118]: I guess this is another case where `impl<T: AsRef<str>> TryFrom<T>` wouldn't work due to existing blanket implementations of `From<T>`?

Comment (roeap) [reply kernel/src/actions/mod.rs:118]: That is correct :)

File: kernel/src/actions/mod.rs (hunk 2 above)
Comment (roeap) [reply kernel/src/actions/mod.rs]: done - i took the liberty of moving the `column_mapping` module as well into a `features` module.

File: kernel/src/snapshot.rs [hunk 4]
```diff
@@ -404,8 +404,8 @@ mod tests {
         let expected = Protocol {
             min_reader_version: 3,
             min_writer_version: 7,
-            reader_features: Some(vec!["deletionVectors".into()]),
-            writer_features: Some(vec!["deletionVectors".into()]),
+            reader_features: Some(vec!["deletionVectors".into()].into_iter().collect()),
+            writer_features: Some(vec!["deletionVectors".into()].into_iter().collect()),
```
Comment (nicklan) [inline kernel/src/snapshot.rs:408]: Why this change. I don't think it's needed?

File: kernel/src/snapshot.rs [hunk 5]
```diff
@@ -426,8 +426,8 @@ mod tests {
...
```
Comment (nicklan) [inline kernel/src/snapshot.rs:430]: ditto, I don't think we need this

**Review by nicklan — CHANGES_REQUESTED** [review]
looks great, other than one (i think) unneeded change

File: kernel/src/actions/mod.rs [hunk 6]
```diff
@@ -112,6 +117,20 @@ impl Protocol {
         data.extract(get_log_schema().project(&[PROTOCOL_NAME])?, &mut visitor)?;
         Ok(visitor.protocol)
     }
+
+    pub fn has_reader_feature(&self, feature: &ReaderFeatures) -> bool {
+        self.reader_features
+            .as_ref()
+            .map(|features| features.iter().any(|f| f == feature.as_ref()))
+            .unwrap_or_default()
```
Comment (scovich) [inline kernel/src/actions/mod.rs:125]:
```suggestion
            .is_some_and(|features| features.iter().any(|f| f == feature.as_ref()))
```
(again below)

File: kernel/src/actions/visitors.rs [hunk 7]
```diff
@@ -365,8 +365,8 @@ mod tests {
         let expected = Protocol {
             min_reader_version: 3,
             min_writer_version: 7,
-            reader_features: Some(vec!["deletionVectors".into()]),
-            writer_features: Some(vec!["deletionVectors".into()]),
+            reader_features: Some(vec!["deletionVectors".into()].into_iter().collect()),
+            writer_features: Some(vec!["deletionVectors".into()].into_iter().collect()),
```
Comment (scovich) [inline kernel/src/actions/visitors.rs:369]: No longer needed, since we're using Vec again?

File: kernel/src/features/mod.rs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+impl TryFrom<&str> for ReaderFeatures {
+    type Error = Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let val = match value {
+            "columnMapping" => ReaderFeatures::ColumnMapping,
+            "deletionVectors" => ReaderFeatures::DeletionVectors,
+            "timestampNtz" => ReaderFeatures::TimestampWithoutTimezone,
+            "v2Checkpoint" => ReaderFeatures::V2Checkpoint,
```
Comment (scovich) [inline kernel/src/features/mod.rs:42]: This replication of magic (string) constants makes me nervous -- no compile-time protection against typos or omissions. Do we have complete round trip unit tests which could somehow catch both present and future typos/incompleteness in this list?

Comment (scovich) [reply kernel/src/features/mod.rs:42]: Actually, we might want to consider using the strum crate, which has derive macros for both [AsRef](https://docs.rs/strum/latest/strum/derive.AsRefStr.html) and [FromStr+TryFrom](https://docs.rs/strum/latest/strum/derive.EnumString.html).

**Comment by nicklan** [comment]
@roeap do you have time to pick this back up? I think it would be nice to get in. I can take it over if you don't. 

Thanks!

**Comment by roeap** [comment]
@nicklan - yes, I'll pick it up ...

File: kernel/src/features/mod.rs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_roundtrip_reader_features() {
```
Comment (scovich) [inline kernel/src/features/mod.rs:183]: Round trip tests are good, but they can only validate the cases we remember to add test cases for... is there a way to future-proof them? (see other comment about maybe deriving both directions, so they can't get out of sync)

File: kernel/src/lib.rs
[the remark this answers is not in the document]
Comment (roeap) [reply kernel/src/lib.rs:1]: yeah, probably since i merged some 60 commits :) - will fix that later today.

File: kernel/src/features/mod.rs (hunk 9 above)
Comment (roeap) [reply kernel/src/features/mod.rs:183]: yeah - so we had similar discussions when we created to config system for object_store. Essentially the thing we wanted to at least centralize the logic, so that we could only break in one place. unfortunately we did not fond a satisfying way to safeguard against typos et. al.

Initially I thought we would need to have multiple strings on the parsing side, which is why I did not use strum. however here that does not apply. WE will need to consider this when we are parsing the configuration though. IIRC for append only both `appendOnly` and `delta.appendOnly` are permissible keys.

So here I plan on adding strum as well, but we'll have the same question again when we do configuration parsing, where I believe we will have to live with custom implementing both ways ... but happy to find a way here :).

Comment (roeap) [reply kernel/src/features/mod.rs:183]: actually, I might be wrong and strum supports multiple string variations on parsing.

Comment (roeap) [reply kernel/src/features/mod.rs:183]: it does :)

**Comment by roeap** [comment]
@scovich @nicklan - I think all comments are addressed. needed to increase the tokio constraint to make things compile for some reason.

also thought about if it makes sense to just make the parsing case invariant, which strum also supports :).

File: kernel/src/features/mod.rs (hunk 9 above)
Comment (scovich) [reply kernel/src/features/mod.rs:70]: Does it also have a way to iterate over all variants, so the unit test stays up to date automatically as new features land?

**Review by scovich — APPROVED** [review]
Nice!

File: kernel/src/features/mod.rs (hunk 9 above)
Comment (roeap) [reply kernel/src/features/mod.rs:92]: Good call - several actually. I opted for exposing the string representation as `ReaderFeatures::VARIANTS` via `VaraintNames` and then include checks that we cover all variants in the roundtrip test. I thought this might be best, as we still have a way to see the variant names in code in case there are typos in the enum and tests also should be forced to stay in sync.

**Review by nicklan — APPROVED** [review]
lgtm! thanks
